### ISSUE TITLE
CRUD cache not cleared when object meta is directly updated

### ISSUE THREAD

**Issue by ragulka** [woocommerce/woocommerce#15274]
### EXPLANATION OF THE ISSUE

I'm not sure if this has popped up before (quick search for crud + cache did not reveal any similar issues), but if it has, feel free to direct me to the correct issue.

It looks like WC_Data meta cache is not cleared if the underlying meta is updated/changed directly with `update_post_meta()`. I ran into this recently with Tab Manager, where on the product edit screen, WC will at some point read the product meta. After that, we try to update a meta key for the product directly with `update_post_meta()`. Yet again later, another piece of code is updating the product meta with `$product->update_meta_data()`. Now, `update_meta_data()` will loop over existing meta that has been fetched earlier and cached as well. This means it will overwrite the meta set with `update_post_meta()`, as that particular meta key was not in the cache.

A dumbed-down version of what's happening can be replicated with the code below:
```php
update_post_meta( $id, ‘_some_meta_key’, 'old_value' );

$product = wc_get_product( $id );
$product->get_meta_data(); // this will return 'old_value' for _some_meta_key

update_post_meta( $id, ‘_some_meta_key’, 'new_value' );

$product = wc_get_product( $id ); // re-instantiate the product

$product->get_meta_data(); // this will return a cached version ( 'old_value' ) of _some_meta_key
```

I understand that starting with 3.0 one of the goals is to move updating any product data to the CRUD methods, but there are plenty of 3rd party plugins and custom code that's still simply using the meta methods directly. This could very well lead to issues with overwriting data that should not be overwritten, or simply getting outdated data if not force-read before.

Correct me if I'm wrong but to me it seems that the object meta cache set in `WC_Data::read_meta_data` should be cleared when any meta key associated with the object is added, updated or deleted?

Should be doable by hooking into the appropriate metadata hooks and calling `wp_cache_delete()` for the appropriate cache key.

Test was done with WC 3.0.7.

**Comment by claudiulodro** [comment]
I wasn't able to reproduce with your sample code. I was able to reproduce your problem with the following unit test:

```
/**
 * Tests that the meta data cache gets flushed when update_post_meta updates the object's meta.
 * @see https://github.com/woocommerce/woocommerce/issues/15274
 */
function test_get_meta_data_after_update_post_meta() {
	$object  = new WC_Product;
	$object->save();
	update_post_meta( $object->get_id(), 'test_meta_key', 'val1' );

	$product = wc_get_product( $object->get_id() );
	$product->update_meta_data( 'test_meta_key_2', 'blah1' );
	update_post_meta( $object->get_id(), 'test_meta_key', 'val2' );
	$product->update_meta_data( 'test_meta_key_2', 'blah2' );
	$product->save();

	$product = wc_get_product( $object->get_id() );
	$metas = $product->get_meta_data();
	$this->assertEquals( 'val2', $metas[0]->value );
}
```

Let me know if this isn't the essence of your problem. 

I don't see a good solution for this, and clearing the cache won't solve it because clearing the wp cache can't update the object's internal meta data cache in `$product->meta_data`. It seems to me like this is the equivalent of a "merge conflict": the data in the db has updated one way and the data in the object has updated a different way and now it has to be resolved.

@mikejolley @claudiosanches any ideas?

**Comment by claudiosanches** [comment]
I can feel the problem and where this hurts... Still we had so many feedback to start a CRUD and let me more easy for update products and orders, and of course, to migrate to a custom database, so not sure if we should support anything updated on `update_post_meta()`, even because we may move products to a custom database too, so better make people already always use CRUD.

**Comment by ragulka** [comment]
> not sure if we should support anything updated on update_post_meta(), even because we may move products to a custom database too, so better make people already always use CRUD.

@claudiosanches I think as long as WC is still using CPT as it's data store, it needs to support it and not break under normal usage. "Normal usage" being here the WordPress Way.

Not everybody uses the CRUD methods to update product meta yet, especially since it's currently rather expensive (updates meta keys that have not changed) and can have unexpected side effects. Because of the caching issue and https://github.com/woocommerce/woocommerce/pull/14751, we've moved away from using the CRUD methods for updating product meta in at least one of our extensions.

_Any_ plugin, theme or customization can potentially break WC and its extensions by simply updating the post meta directly.

**Comment by mikejolley** [comment]
Submitting unit tests as PRs for issues like this are welcome FWIW.

**Comment by crodas** [comment]
> Not everybody uses the CRUD methods to update product meta yet, especially since it's currently rather expensive (updates meta keys that have not changed) and can have unexpected side effects. Because of the caching issue and #14751, we've moved away from using the CRUD methods for updating product meta in at least one of our extensions.

@ragulka  Is there anything wrong with my PR?

**Comment by crodas** [comment]
@claudiulodro I fixed that problem in #14751, and I also improved it by saving only changes (not everything as it is now).

**Comment by ragulka** [comment]
> Is there anything wrong with my PR?

@crodas Sorry, I was referring to the issue your PR tries to solve. Nothing wrong with the PR itself :)
